Tolerate parse errors in terms

Avi-D-coder opened the pull request:
This is the beginning of my effort to produce useful information for NLS in the presence of errors.

We now type check and parse with errors.

yannham commented on src/cache.rs:
```diff
@@ -290,30 +290,57 @@ impl Cache {
+    /// entry has already been parsed.
+    ///
+    /// The cache will be populated even if there are parse errors.
+    pub fn parse_tolerate_errors(&mut self, file_id: FileId) -> Result<CacheOp<()>, ParseErrors> {
```
The same remark as my other comment below (sorry, I did a non-linear review :sweat_smile: ): I dont think `ParseErrors` should go in the `Err` part of the result. IMO we should differentiate between recoverable parsing errors and fatal parsing errors in the return type.

yannham reviewed (commented):
This is the way to go, but I left a few comments about the API. In practice though, I have a mixed experience. If I use the following chunk:

```
let y = (.) in
let x = {foo = 1, bar = 2, baz = 3} in
(.) + x.bar == 3
```

I do get two different parsing errors, that seems to work. However, I'm still unable to use other LSP functions not involved in the problem like go to definition for `x` or hover on other terms like `3`, and that even if I remove the second parsing error. Do you have an idea about what's happening? @ysndr maybe this is not sufficient to make the linearization work?

Avi-D-coder commented on src/cache.rs:
```diff
@@ -331,22 +361,43 @@ impl Cache {
         file_id: FileId,
         format: InputFormat,
-    ) -> Result<RichTerm, ParseError> {
+    ) -> Result<RichTerm, (Option<RichTerm>, ParseErrors)> {
```
I tried a few different signatures and picked this one since it was least invasive.
I'll change it back to `Result<(RichTerm, ParseErrors), ParseError>`.

yannham replied:
Well, unless you have experienced that it is much more invasive, in which case that would be a valid reason not to. But it seems that the change would be limited to those few caches functions, and wouldn't bubble up that much, or would they?

Avi-D-coder replied:
Yeah, changing it is a good idea.

Avi-D-coder commented:
> This is the way to go, but I left a few comments about the API. In practice though, I have a mixed experience. If I use the following chunk:
> 
> ```
> let y = (.) in
> let x = {foo = 1, bar = 2, baz = 3} in
> (.) + x.bar == 3
> ```
> 
> I do get two different parsing errors, that seems to work. However, I'm still unable to use other LSP functions not involved in the problem like go to definition for `x` or hover on other terms like `3`, and that even if I remove the second parsing error. Do you have an idea about what's happening? @ysndr maybe this is not sufficient to make the linearization work?

I have been trying to narrow lexical scope of the errors, but the grammar keeps failing.

Avi-D-coder commented on src/typecheck/mod.rs:
```diff
@@ -552,8 +552,8 @@ fn type_check_<S, E>(
-        // null is inferred to be of type Dyn
-        Term::Null => unify(state, strict, ty, mk_typewrapper::dynamic())
+        // null and parse errors are inferred to be of type Dyn
+        Term::Null | Term::ParseError => unify(state, strict, ty, mk_typewrapper::dynamic())
```
This improves things a good deal, but there are a number caching issues.
I'm working on invalidating the cache after changes to parse errors.

Avi-D-coder commented:
@yannham I addressed the reviews, and I added non fatal parse errors to the cache. Caching parse errors prevents us losing them when a file is first parsed from a caller that ignores non fatal errors.

yannham commented:
Haven't reviewed the new code yet, but just tested it and it seems to work :tada: !

yannham commented on src/cache.rs:
```diff
@@ -62,7 +62,7 @@ pub struct Cache {
     imports: HashMap<FileId, HashSet<FileId>>,
     /// The table storing parsed terms corresponding to the entries of the file database.
-    terms: HashMap<FileId, (RichTerm, EntryState)>,
+    terms: HashMap<FileId, (RichTerm, EntryState, ParseErrors)>,
```
Nitpicking: it might be worth making a separate struct `CacheEntry` or something, for readability and ease of pattern matching.

Avi-D-coder replied:
Refactored
